backend/accounting-service/src/main/kotlin/services/projects/v2/ProjectService.kt

```diff
@@ -1245,6 +1249,267 @@ class ProjectService(
+                    HttpStatusCode.BadRequest
+                )
+            }
+        }
+    }
+
+    suspend fun updateInviteLinkGroupAssignment(
+        actorAndProject: ActorAndProject,
+        request: ProjectsUpdateInviteLinkGroupAssignmentRequest,
+        ctx: DBContext = db
+    ) {
+        val project = actorAndProject.requireProject()
+
+        ctx.withSession { session ->
+            requireAdmin(actorAndProject.actor, listOf(project), session)
+
+            session.sendPreparedStatement(
+                {
+                    setParameter("project", project)
+                    setParameter("token", request.token)
+                    setParameter("groups", request.groups)
+                },
+                """
+                    with new (group_id, token) as (
+                        select id, cast(:token as uuid) from project.groups where project = :project and id in (select unnest(cast(:groups as text[])))
+                    ), changed as (
+                        insert into project.invite_link_group_assignments (group_id, link_token)
+                        select * from new
+                        on conflict do nothing
+                    ) delete from project.invite_link_group_assignments
+                    where link_token = cast(:token as uuid) and (group_id, link_token) not in (select group_id, token from new)
+                """
+            )
+        }
+    }
+
+    suspend fun acceptInviteLink(
+        actorAndProject: ActorAndProject,
+        request: ProjectsAcceptInviteLinkRequest,
+        ctx: DBContext = db
+    ) : ProjectsAcceptInviteLinkResponse {
+        data class ProjectInviteLinkAssignment(
+            val piActorAndProject: ActorAndProject,
+            val role: ProjectRole,
+            val groups: List<String>
+        )
+
+        return ctx.withSession { session ->
+            val projectAssignment = session.sendPreparedStatement(
+                {
+                    setParameter("token", request.token)
+                },
+                """
+                    select m.project_id as project, m.username as pi, l.role_assignment as role, a.group_id as group_id
+                    from project.project_members m
+                    inner join project.invite_links l on
+                        now() < expires and
+                        token = cast(:token as uuid) and
+                        l.project_id = m.project_id
+                    left join project.invite_link_group_assignments a on
+                        link_token = cast(:token as uuid)
+                    where
+                        m.role = 'PI';
+                """
+            ).rows.groupBy { it.getString("project") }.map { entry ->
+                val project = entry.key
+                val pi = entry.value.first().getString("pi")
+                val role = entry.value.first().getString("role")
+
+                if (project == null || pi == null || role == null) {
```

Comment:
This will never happen since `first()` will crash the request with an internal server error instead. See https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/first.html#exceptions